### PR TITLE
Provenance

### PR DESCRIPTION
**Pull request by jdoliner**

### CODE REVIEW

File: src/server/pachyderm_test.go [hunk 1]
```diff
@@ -1247,6 +1254,67 @@ func TestAssets(t *testing.T) {
...
+	require.NoError(t, c.FinishCommit(aRepo, commit2.ID))
+	commitInfos, err := c.FlushCommit([]*pfsclient.Commit{client.NewCommit(aRepo, commit2.ID)}, nil)
+	require.NoError(t, err)
+	require.Equal(t, 2, len(commitInfos))
+
+	// There should only be 2 commits on cRepo
+	commitInfos, err = c.ListCommit([]string{cPipeline}, nil, pfsclient.CommitType_COMMIT_TYPE_READ, false, false)
+	require.NoError(t, err)
+	require.Equal(t, 2, len(commitInfos))
+}
```
Comment (derekchiang) [inline src/server/pachyderm_test.go]: Could you make pipeline C use data from both A and B, and make sure that pipeline C is actually producing the expected output?

File: src/server/pfs/drive/driver.go [hunk 2]
```diff
@@ -705,26 +741,79 @@ func (d *driver) inspectRepo(repo *pfs.Repo, shards map[uint64]bool) (*pfs.RepoI
...
 	}
+	provenance, err := d.repoProvenance(repo, shards)
+	if err != nil {
+		return nil, err
+	}
+	result.Provenance = provenance
+	return result, nil
+}
+
+func (d *driver) recurseProvenance(commit *pfs.Commit, repoSet map[string]bool,
```
Comment (derekchiang) [inline src/server/pfs/drive/driver.go]: What does this function do?  Some comments would help

File: src/server/pfs/reduce.go [hunk 3]
```diff
@@ -16,6 +16,7 @@ func ReduceRepoInfos(repoInfos []*pfs.RepoInfo) []*pfs.RepoInfo {
 			continue
 		}
 		reducedRepoInfo.SizeBytes += repoInfo.SizeBytes
+		reducedRepoInfo.Provenance = repoInfo.Provenance
```
Comment (derekchiang) [inline src/server/pfs/reduce.go]: Is it an invariant that the `Provenance` returned by each server should be the same?  If so, can we assert that?

File: src/server/pfs/reduce.go [hunk 4]
```diff
@@ -37,6 +38,7 @@ func ReduceCommitInfos(commitInfos []*pfs.CommitInfo) []*pfs.CommitInfo {
 			reducedCommitInfo.CommitType = pfs.CommitType_COMMIT_TYPE_WRITE
 		}
 		reducedCommitInfo.SizeBytes += commitInfo.SizeBytes
+		reducedCommitInfo.Provenance = commitInfo.Provenance
```
Comment (derekchiang) [inline src/server/pfs/reduce.go]: Same comment as above

File: src/server/pfs/reduce.go (hunk 3 above)
Comment (jdoliner) [inline src/server/pfs/reduce.go]: It is an invariant. Asserting it gets a bit tricky though. For one reduce functions don't return errors so we'd have to add that and handle them. For a second there's no guarantee about the order of the `Provenance` although it should in theory be the same. We'd probably want to compare them as sets. Considering that we're going to be getting rid of this code when we do the pfs refactor my vote would be to ignore for now. I can implement it though if that makes you uneasy.

Comment (derekchiang) [inline src/server/pfs/reduce.go]: OK that seems reasonable.  Ignore it then.

**Comment by derekchiang** [comment]
LGTM

**Pull request merged**